[Gan Hong Yao] iP

ganhongyao opened the pull request:
## Duke

> "If you fail to plan, you are planning to fail." - Benjamin Franklin

Try out Duke **today**!

### Features
- Add tasks
- Mark tasks as done
- Delete tasks
- Save tasks to local file
- Find tasks by name or date

### Getting started
To get started with Duke, it's as easy as 1️⃣ 2️⃣ 3️⃣ !
1. Download the latest release from [here](https://github.com/ganhongyao/ip/releases).
2. Double-click to open it.
3. _Happy Duking!_

### To-do
- [x] Add GUI
- [ ] Add more unit tests (any help is deeply appreciated 😆 )

Unit test methods should adopt the following naming convention `whatIsBeingTested_descriptionOfTestInputs_expectedOutcome`. 

Example of a test method:
```java
@Test
public void execute_nonIntegerArgument_exceptionThrown() {
    try {
        new DeleteTaskCommand("blah")
                .execute(new TaskManager(), new Storage("./data/tasks.txt"));
        fail();
    } catch (DukeException e) {
        assertEquals("Invalid task number.", e.getMessage());
    }
}
```

If you find any bugs, please ~~fix it~~ report it to me!

ganhongyao committed:
Add Level 1 - Greet, Echo, Exit

ganhongyao committed:
Merge pull request #1 from ganhongyao/level-1

Add Level 1 - Greet, Echo, Exit (#1)

ganhongyao committed:
Add Level 2: Add, List (#2)

* Change indent method in ResponsePrettifier to indent all lines

* Complete Level 2

ganhongyao committed:
Add Level 3: Mark as Done (#3)

ganhongyao committed:
Add Level 4: ToDos, Events, Deadlines (#4)

* Add ToDo class

* Add Deadline class

* Add Event class

* Make Task class abstract

* Add more descriptive display messages

ganhongyao committed:
Improve parsing of command arguments (#5)

* Improve parsing of arguments

* Print extra new line after response

ganhongyao committed:
Add Text UI Testing (#6)

* Update .gitignore

* Update test files

ganhongyao committed:
Add Level 5: Handle Errors (#7)

* Add DukeException

* Wrap command handling in a try-catch block

ganhongyao committed:
Add Level 6: Delete (#8)

* Use DukeException in TaskManager

* Add delete task functionality

* Update test files

* Add test case for IndexOutOfBounds

ganhongyao committed:
Add Enums (#9)

* Add enum for commands

* Reorder cases according to enum declaration

ganhongyao committed:
Improve regex matching (#10)

* Split by any number of any whitespace characters instead of single space

* Add limit argument to split

ganhongyao committed:
Add text representations for tasks

ganhongyao committed:
Add parsing of text to Tasks

ganhongyao committed:
Add Storage class

ganhongyao committed:
Ensure tasks are loaded and saved correctly

ganhongyao committed:
Simplify Storage logic

ganhongyao committed:
Update .gitignore to ignore data directory

ganhongyao committed:
Standardise error handling within Storage

ganhongyao committed:
Merge pull request #12 from ganhongyao/branch-Level-7

Add Level 7: Save

ganhongyao committed:
Add support for date and times

ganhongyao committed:
Update tests after adding date time support

ganhongyao committed:
Allow users to list tasks on certain date

ganhongyao committed:
Merge pull request #13 from ganhongyao/branch-Level-8

Add Level 8: Dates and Times

ganhongyao committed:
Add more OOP (#14)

* Refactor ResponsePrettifier to Ui

* Extract Command classes and add Parser

ganhongyao committed:
Organise into packages (#15)

* Add duke as top level package

* Split classes into relevant packages

* Remove unused variables

ganhongyao committed:
Update runtest.sh after adding packages (#16)

ganhongyao committed:
Add JUnit tests (#17)

* Add Gradle support

* Add checkstyle config but comment out checkstyle task

* Add ParserTest

* Add tests for CompleteTaskCommand and DeleteTaskCommand

ganhongyao committed:
Add docs

ganhongyao committed:
Remove unnecessary use of 'this'

ganhongyao committed:
Add ability to filter tasks by name

ganhongyao committed:
Merge pull request #18 from ganhongyao/branch-A-JavaDoc

Add documentation

ganhongyao committed:
Merge pull request #19 from ganhongyao/branch-A-CodingStandard

Remove unnecessary use of 'this' keyword

ganhongyao committed:
Merge pull request #20 from ganhongyao/branch-Level-9

Add Level 9: Find

ganhongyao committed:
Refactor Storage and Parser (#21)

* Make saveTasks take in TaskManager parameter instead of String paramter

* Rename Parser to CommandParser

* Shift task parsing responsibility to TaskParser instead of individual tasks

* Add more docs for Task class

* Reorder methods in TaskManager

ganhongyao committed:
Fix checkstyle violations (#22)

* Uncomment checkstyle Gradle task

* Fix checkstyle violations

ganhongyao committed:
Update CommandParser JUnit test (#23)

ganhongyao committed:
Add GUI

ganhongyao committed:
Refactor responses from Duke

ganhongyao committed:
Set common entrypoint

ganhongyao committed:
Customise GUI

ganhongyao committed:
Add documentation for DukeBadResponse

ganhongyao committed:
Merge pull request #24 from ganhongyao/branch-Level-10

Add GUI

ganhongyao committed:
Modify Find function to take in any number of search strings (#25)

guangxuan commented on src/main/java/duke/parser/CommandParser.java:
```diff
@@ -0,0 +1,49 @@
+        final String commandWord = trimmedCommand.split("\\s+")[0];
+        final String commandArguments = trimmedCommand.replace(commandWord, "").trim();
+
+        switch (commandWord) {
```
Consider using an enum instead of String

guangxuan commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public String toText() {
+        String[] props = new String[]{"T", super.getStatusIcon(), super.getName()};
```
```suggestion
        String[] props = {"T", super.getStatusIcon(), super.getName()};
```

ganhongyao replied:
Done.

guangxuan commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Returns whether the task is marked as completed.
+     */
+    public boolean checkTaskDone() {
```
```suggestion
    public boolean isDone() {
```

YoYoCiti replied:
I agree with this change as well

ganhongyao replied:
Thanks for the suggestion. Have changed it! 😄

guangxuan reviewed (commented):
LGTM. Very comprehensive analysis of exceptions, commands, and even Parsers!
There is not much to be said about Code Quality. Good job!

YoYoCiti commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+        this.isDone = isDone;
+    }
+
+    protected Task markAsDone() {
```
Maybe method name should be setDone? As coding standard said setter methods for boolean variables should be of that form.

ganhongyao replied:
I personally think `setDone` would be a more appropriate name for an actual setter method i.e. 
```java
protected Task setDone(boolean isDone) {
    this.isDone = isDone;
}
```
, which may be implemented in future. Hence I chose the name `markAsDone` so that we don't confuse the methods.

Thanks for the suggestion though!

YoYoCiti commented on src/main/java/duke/parser/CommandParser.java:
```diff
@@ -0,0 +1,49 @@
+public class CommandParser {
+    /**
+     * Parses the command as well as any arguments entered by the user.
+     * @param fullCommand the full command entered by the user
```
I think there should be an empty line between the description and parameter section? Noticed this is a recurring issue

ganhongyao replied:
You are right. Thanks for pointing this out to me! Have fixed this. 😄

YoYoCiti commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,208 @@
+
+    /**
+     * Deletes a <code>Task</code> from the list.
+     * @param taskNumber the number of the task to be deleted
```
I think a punctuation is required behind the parameter description?

ganhongyao replied:
Yup you're right. Have fixed this!

YoYoCiti reviewed (commented):
Overall, the code was very easy and pleasant to read. Only picked out extremely nitpicky and minor coding standard violations. Thanks for the read :smile:

ganhongyao committed:
Add more style customisation (#26)

* Change font to YuseiMagic

* Change dialog colors

* Increase width of MainWindow

* Reorder variable declaration

* Add app icon

ganhongyao committed:
Fix coding standard violations (#27)

* Modify array initialisation

* Rename getter method to sound more like boolean

* Fix javadocs

ganhongyao committed:
Add more javadocs fixes (#28)

ganhongyao committed:
Specify minHeight attribute on Label instead of root (#29)

ganhongyao committed:
Fix Qoobee image

ganhongyao committed:
Fix background image not covering full window when scrollbar not present

ganhongyao committed:
Merge pull request #30 from ganhongyao/fix-images

Fix Qoobee image and background image

ganhongyao committed:
Assert that Task to be added is not null

TaskManager::addTask assumes that the Task argument is a
valid Task and adds it directly to the taskList.

This makes null a valid argument and may cause unexpected exceptions
when we operate on it.

Let's add an assertion at the start of TaskManager::addTask that the
Task to be added cannot be null.

ganhongyao committed:
Merge pull request #31 from ganhongyao/branch-A-Assertions

Assert that Task to be added is not null

ganhongyao committed:
Extract formatting of tasks as list into function

The overloaded methods of TaskManager::list filter the tasks (if
applicable), get the string representations of the individual tasks in
an array of strings and then join the strings to form the returned
string.

This common behaviour causes code duplication.

Extracting this behaviour into a function allows us to reduce code
duplication by calling this function from each list method.

Let's create a method formatTasksAsList to address the code duplication.

ganhongyao committed:
Merge pull request #32 from ganhongyao/branch-A-CodeQuality

Extract formatting of tasks as list into function

ganhongyao committed:
Use streams in filter methods

ganhongyao committed:
Merge pull request #33 from ganhongyao/branch-A-Streams

Use streams in filter methods

ganhongyao committed:
Add CI config

ganhongyao committed:
Merge pull request #34 from ganhongyao/branch-A-CI

Add CI configuration

ganhongyao committed:
Add command to allow users to load tasks from another file

ganhongyao committed:
Add support for config file

ganhongyao committed:
Add Javadocs

ganhongyao committed:
Merge pull request #35 from ganhongyao/branch-C-FlexibleDataSource

Add more flexibility for data source

ganhongyao committed:
Add exceptions subclasses (#36)

* Add InvalidCommandUsageException

* Add TaskParseException

ganhongyao committed:
Fix use of magic literals (#37)

ganhongyao committed:
Update User Guide

ganhongyao committed:
Add screenshot to docs

ganhongyao committed:
Fix JavaFX dependency (#39)
